### ISSUE TITLE
Orthonormalization Issues in eigs

### ISSUE THREAD

**Issue by lbalicki** [pymor/pymor#1067]
When running the following code I get `AccuracyError: result not orthogonal (max err=0.0013971279933609176)` after a few steps (also happens for slightly different arguments of the `eigs` function):

```python
import numpy as np
import scipy.sparse as sps

from pymor.algorithms.eigs import eigs
from pymor.models.iosys import LTIModel

k = 50
n = 2 * k + 1

A = sps.diags(
    [(n - 1) * [(n - 1)**2], n * [-2 * (n - 1)**2], (n - 1) * [(n - 1)**2]],
    [-1, 0, 1],
    format='lil',
)
A[0, 0] = A[-1, -1] = -2 * n * (n - 1)
A[0, 1] = A[-1, -2] = 2 * (n - 1)**2
A = A.tocsc()

B = np.zeros((n, 2))
B[:, 0] = 1
B[0, 1] = 2 * (n - 1)

C = np.zeros((3, n))
C[0, 0] = C[1, k] = C[2, -1] = 1

fom = LTIModel.from_matrices(A, B, C)

eigs(fom.A, k=20, which='LR')
```

The issue is that in `eigs` when adding vectors to the Arnoldi-basis, only the most recently computed basis vector is orthonormalized against the rest of the basis via `gram_schmidt(V, return_R=True, atol=0, rtol=0, offset=len(V) - 1, copy=False)`. When removing the offset in `eigs`, i.e. orthonormalizing the entire basis in every step of the iteration via `gram_schmidt(V, return_R=True, atol=0, rtol=0, copy=False)` everything works well. So it seems that the errors in the orthonormalization process propagate themselves and finally make it impossible to orthonormalize a single new vector against the other vectors in the VectorArray `V`.

It would be nice, to have an option in `gram_schmidt` that allows for reorthonormalizing the entire `VectorArray` even if `offset` is not zero. Maybe something like a `reiterate_offset` option would be good, which indicates the offset when reiterating? Or a boolean option `reiterate_full` which sets the offset to zero when reiterating?

**Comment by pmli** [comment]
I found that increasing the reiteration threshold by adding `reiteration_threshold=0.5` makes it work, so just adding one reorthogonalization for the last vector seems to be enough.

There is the "twice is enough" rule of thumb that I had been thinking about adding to `gram_schmidt` and this example looks like good motivation for this. [SLEPc technical report](https://pdfs.semanticscholar.org/8821/ecb171659e11f334518d1654d7b1aa180a08.pdf) has a nice discussion about this and different reorthogonalization criteria. As far as I understood, the default behavior they chose is reorthonormalizing at least once and then checking the criterion `gram_schmidt` uses with `reiteration_threshold=1/sqrt(2)`. So, maybe adding something like `reiteration_min_num` to `gram_schmidt` to set the minimum number of reorthonormalizations would be good.

**Comment by lbalicki** [comment]
I did some testing and increasing `reiteration_threshold` to `0.5` gives good results for every case that I have tried. The way I understood the SLEPc technical report is that (as the default option) they chose to implement orthogonalization just like `gram_schmidt` in pyMOR, but with a higher value for `reiteration_threshold` and more reorthogonalization steps. Always reiterating seems unnecessary to me since it is expensive and not all algorithms need this. So overall, for now I don't think it will be necessary to add any extra options to `gram_schmidt`, but rather I would suggest we increase the default value for `reiteration_thresold`. Is there a reason it is at `0.1` right now or would it make more sense to change the threshold just in the `gram_schmidt` function calls in the `eigs` implementation?

**Comment by pmli** [comment]
I'm not sure about changing or not changing the default values in `gram_schmidt`, but it would be a bit strange if only `eigs` uses a different value. I guess another option would be adding parameters to `eigs` which are passed to `gram_schmidt`.

@sdrave What do you think?

**Comment by sdrave** [comment]
> Is there a reason it is at 0.1 right now or would it make more sense to change the threshold just in the gram_schmidt function calls in the eigs implementation?

I think it is just some arbitrary number. Maybe @andreasbuhr can comment, as he originally introduced re-orthogonalization in the algorithm.

In general I would be open to change the default, however, I guess 0.5 would be just another random number which makes some example work. Changing the default might also increase the time required for offline-online decomposition of the standard RB error estimator. Maybe it is possible to find a better criterion when to re-iterate?

**Comment by pmli** [comment]
It looks like [ARPACK](https://github.com/scipy/scipy/blob/d3416c7675d45c81bd19dfed36267832ddc453b9/scipy/sparse/linalg/eigen/arpack/ARPACK/SRC/dnaitr.f#L642) uses 0.717 (which is a bit bigger than `1/sqrt(2) = 0.7071...`).

There are other criteria, some also mentioned in the SLEPc technical report, but I'm not sure if any is universally better.

**Comment by andreasbuhr** [comment]
Hi all. Yes, 0.1 was just some arbitrary number. Accepting a little error to save an iteration now and then.

Re-iterating twice is not enough. You can find a short four-page analysis in my PhD thesis, Appendix C. It is a self-contained appendix, page 119 to 122 (pdf-page 139 to 142) of the thesis which can be found here: https://arxiv.org/pdf/1908.02074 . See especially Figure C.1 which shows that errors do shoot up with two iterations.

I think a reiteration_threshold default of 0.717 would be a good idea. I remember I often used 0.99, just to be sure.

**Comment by sdrave** [comment]
I would agree to increase the threshold and make it configurable for the residual reductors (which usually don't require too much precision). However, I'm not so sure about 0.717 as it seems to be just another random number (maybe even a typo). I would agree to 0.707, as there seems to be some theoretical evidence that it might be a good choice, at least in some situations.

**Comment by sdrave** [comment]
I just checked that for instance the runtime of

```
thermalblock.py 2 3 3 60 --grid=300
```

does not substantially increase when `reiteration_tol` is set to 9e-1. This makes sense: most of the offline time is spent in orthonormalizing a basis for the MOR residual. The basis vectors quickly become almost linearly dependent so the already for the smaller `reiteration_tol` almost every vector is reorthonormalized anyway (multiple reiterations do not seem to be necessary in both cases). So I don't see much drawback in choosing the stricter value. I have prepared a PR in #1154.
